### PR TITLE
Introduce program creation from program in `gtest`

### LINKED ISSUE #714: Implement program from program feature in `gtest`

### PR DESCRIPTION
**Pull request by SabaunT**
Resolves #714 .

Release notes: now `gtest` developers can test program from program creation logic in their smart contracts. An example is shown in `tests/program-factory` crate.

### CODE REVIEW

File: gtest/src/system.rs [hunk 1]
```diff
@@ -62,10 +66,53 @@ impl System {
...
+    /// function, developer should provide to the function "child's" code hash. Code for that
+    /// code hash must be in storage at the time of the function call. So this method stores
+    /// the code in storage.
+    pub fn submit_code<P: AsRef<Path>>(&self, code_path: P) -> CodeHash {
+        let path = env::current_dir()
+            .expect("Unable to get root directory of the project")
+            .join(code_path)
+            .clean();
+
+        let code = fs::read(&path).unwrap_or_else(|_| panic!("Failed to read file {:?}", path));
```
[the remark this answers is not in the document]
Comment (SabaunT) [reply gtest/src/system.rs:115]: Clippy doesn't like that: if you have a function call inside "unwrapping" action, you'd better use `unwrap_or_else`.

File: gtest/src/manager.rs [hunk 2]
```diff
@@ -538,11 +557,34 @@ impl JournalHandler for ExtManager {
...
-        &mut self,
-        _code_hash: CodeHash,
-        _candidates: Vec<(ProgramId, MessageId)>,
-    ) {
-        // todo!() #714
+    fn store_new_programs(&mut self, code_hash: CodeHash, candidates: Vec<(ProgramId, MessageId)>) {
+        if let Some(code) = self.codes.get(&code_hash).cloned() {
+            for (candidate_id, init_message_id) in candidates {
+                if !self.actors.contains_key(&candidate_id) {
+                    let candidate = CoreProgram::new(candidate_id, code.clone()).unwrap_or_else(|e|
```
[the remark this answers is not in the document]
Comment (SabaunT) [reply gtest/src/manager.rs:564]: Well, we have a `for` loop, which gives ownership over owned vector to `candidate` each time.

File: gtest/src/program.rs [hunk 3]
```diff
@@ -79,6 +79,10 @@ pub struct Program<'a> {
 }
 
 impl<'a> Program<'a> {
+    pub fn calculate_program_id(code_hash: CodeHash, salt: &[u8]) -> ProgramId {
+        ProgramId::generate(code_hash, salt)
+    }
+
```
Comment (breathx) [inline gtest/src/program.rs:85]: I don't like this method (especially inside the `Program` impl) due to ability to manually set id of the program in `gtest`.

File: gtest/src/system.rs [hunk 4]
```diff
@@ -62,10 +66,53 @@ impl System {
...
+            .actors
+            .get(&id)
+            .and_then(|(actor, _)| match actor {
+                Actor::Dormant => None,
+                _ => Some(Program {
+                    id,
+                    manager: &self.0,
+                }),
+            })
+    }
```
Comment (breathx) [inline gtest/src/system.rs:85]: I would like to keep panic here for api-convenience

File: gtest/src/system.rs [hunk 5]
```diff
@@ -62,10 +66,53 @@ impl System {
...
+        self.0
+            .borrow()
+            .actors
+            .iter()
+            .filter_map(|(&program_id, (actor, _))| match actor {
+                Actor::Initialized(_) => Some(program_id),
+                _ => None,
+            })
+            .collect()
+    }
```
Comment (breathx) [inline gtest/src/system.rs:100]: why do we need this method? and can't we replace it with `system.program_active(id) -> bool` ?

**Review by breathx — CHANGES_REQUESTED** [review]

File: gtest/src/manager.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (shamilsan) [reply gtest/src/manager.rs:564]: Ok but you can remove `cloned()`:

```rust
        if let Some(code) = self.codes.get(&code_hash) {
```

**Review by gshep — APPROVED** [review]
looks good to me!

File: gtest/src/manager.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (shamilsan) [reply gtest/src/manager.rs:564]: Ah, I see, self is mutable below, ok.

File: gtest/src/program.rs (hunk 3 above)
Comment (SabaunT) [reply gtest/src/program.rs:85]: discussed, will be moved as a separate function

File: gtest/src/system.rs (hunk 4 above)
Comment (SabaunT) [reply gtest/src/system.rs:85]: discussed with @LouiseMedova and @breathx . Will remove option

File: gtest/src/system.rs (hunk 5 above)
Comment (SabaunT) [reply gtest/src/system.rs]: Actually needed that for the case, when someone tries to create a duplicate program from program. In [de9c434](https://github.com/gear-tech/gear/pull/755/commits/de9c43448972814931501a63dbbcf731887b7ec4) I have introduced `total_processed` messages counter. Now duplicate check can be done by observing amount of messages processed.

File: gtest/src/program.rs (hunk 3 above)
Comment (breathx) [reply gtest/src/program.rs:85]: Isn't it?

**Review by breathx — APPROVED** [review]
Well done 👍

File: gtest/src/program.rs (hunk 3 above)
Comment (SabaunT) [reply gtest/src/program.rs]: moved in [2045fb6](https://github.com/gear-tech/gear/pull/755/commits/2045fb671b8ff915991b79b3af1874092e28c3d3)

**Pull request merged**